sys/net/application_layer/gcoap/forward_proxy.c

```diff
@@ -0,0 +1,435 @@
+ * directory for more details.
+ */
+
+/**
+ * @{
+ *
+ * @file
+ * @author  Cenk Gündoğan <cenk.guendogan@haw-hamburg.de>
+ */
+
+#include "net/gcoap.h"
+#include "net/gcoap/forward_proxy.h"
+#include "uri_parser.h"
+#include "net/nanocoap/cache.h"
+
+#define ENABLE_DEBUG    0
+#include "debug.h"
+
+typedef struct {
+    int in_use;
+    sock_udp_ep_t ep;
+#if IS_ACTIVE(MODULE_NANOCOAP_CACHE)
+    uint8_t cache_key[CONFIG_NANOCOAP_CACHE_KEY_LENGTH];
+#endif
+} client_ep_t;
+
+static uint8_t proxy_req_buf[CONFIG_GCOAP_PDU_BUF_SIZE];
+static client_ep_t _client_eps[CONFIG_GCOAP_REQ_WAITING_MAX];
+
+static int _request_matcher_forward_proxy(gcoap_listener_t *listener,
+                                          const coap_resource_t **resource,
+                                          const coap_pkt_t *pdu);
+static ssize_t _forward_proxy_handler(coap_pkt_t* pdu, uint8_t *buf,
+                                      size_t len, void *ctx);
+
+const coap_resource_t forward_proxy_resources[] = {
+    { "/", COAP_GET, _forward_proxy_handler, NULL },
+};
+
+gcoap_listener_t forward_proxy_listener = {
+    &forward_proxy_resources[0],
+    ARRAY_SIZE(forward_proxy_resources),
+    NULL,
+    NULL,
+    _request_matcher_forward_proxy
+};
+
+void gcoap_forward_proxy_init(void)
+{
+    gcoap_register_listener(&forward_proxy_listener);
+
+    /* initialize the nanocoap cache operation, if compiled */
+    if (IS_ACTIVE(MODULE_NANOCOAP_CACHE)) {
+        nanocoap_cache_init();
+    }
+}
+
+static int _cache_build_response(nanocoap_cache_entry_t *ce,
+                                 coap_pkt_t *pdu,
+                                 uint8_t *buf,
+                                 size_t len)
+{
+    /* Use the same code from the cached content. Use other header
+     * fields from the incoming request */
+    gcoap_resp_init(pdu, buf, len, ce->response_pkt.hdr->code);
+    /* copy all options and possible payload from the cached response
+     * to the new response */
+    unsigned header_len_req = coap_get_total_hdr_len(pdu);
+    unsigned header_len_cached = coap_get_total_hdr_len(&ce->response_pkt);
+    unsigned opt_payload_len = ce->response_len - header_len_cached;
+
+    memcpy((buf + header_len_req),
+           (ce->response_buf + header_len_cached),
+           opt_payload_len);
+    return header_len_req + opt_payload_len;
+}
+
+static int _cache_lookup_and_process(coap_pkt_t *pdu,
+                                     uint8_t *buf,
+                                     size_t len,
+                                     client_ep_t *cep)
+{
+    (void) cep;
+
+    uint8_t cache_key[SHA256_DIGEST_LENGTH];
+    uint64_t now = xtimer_now_usec64();
+    nanocoap_cache_key_generate(pdu, cache_key);
+    nanocoap_cache_entry_t *ce = nanocoap_cache_key_lookup(cache_key);
+
+    /* cache hit, methods are equal, and cache entry is not stale */
+    if (ce &&
+        (ce->request_method == coap_get_code(pdu)) &&
+        (ce->max_age > now)) {
```

Comment:
Couldn't this already be checked within `nanocoap_cache_key_lookup()`? It would prevent some pitfalls if someone does their own caching implementation and did not think about checking the age of the entry. `nanocoap_cache` includes a timer anyways, so I do not see why this check can't be done internally.

Reply:
With Cache validation in mind, it makes sense to check it outside of `lookup`, so the comment can be ignored.